### PR TITLE
OAuth2

### PR DESCRIPTION
**Pull request by wild-endeavor**
This adds OAuth2 support for Flyte Admin.

* Most of the logic has been implemented in the auth package.  Care was taken, at least attepted, to separate logic that exclusively dealt with oauth into this package.
* The code implements the authorization code flow, which means you'll need a client secret, and there is no implicit flow.
* Most of the core logic is in the `handlers.go` file, which has functions that create handlers (both HTTP and gRPC) that can be attached to existing servers.  See the `serve.go` file to see usage.
* Encrypted cookies are used to store the user's access (oidc) and refresh tokens.  Two other cookies (the CSRF token and the final redirect page) are stored unencypted.
 * Reworked some of the options handling around the base server logic in `serve.go`, specifically around SSL handling.

Docs forthcoming in the main flyte repo.  Also more unit tests and metrics here in another PR after the remaining auth work is done.

### CODE REVIEW

File: pkg/server/cors.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package server
+
+import (
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"github.com/grpc-ecosystem/grpc-gateway/utilities"
+)
+
+const (
+	// At time of writing, the version argument is required to be 1.
```
Comment (katrogan) [inline pkg/server/cors.go]: can you explain why or link to something that does?

Comment (wild-endeavor) [reply pkg/server/cors.go]: added.

File: pkg/auth/handlers.go [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+		http.SetCookie(writer, &csrfCookie)
+
+		state := HashCsrfState(csrfToken)
+		logger.Debugf(ctx, "Setting CSRF state cookie to %s and state to %s\n", csrfToken, state)
+		url := authContext.OAuth2Config().AuthCodeURL(state)
+		fmt.Println(url)
+		queryParams := request.URL.Query()
+		if flowEndRedirectUrl := queryParams.Get(LoginRedirectUrlParameter); flowEndRedirectUrl != "" {
+			redirectCookie := NewRedirectCookie(ctx, flowEndRedirectUrl)
+			if redirectCookie != nil {
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply pkg/auth/handlers.go]: will address with a comment.

[the remark this answers is not in the document]
Comment (wild-endeavor) [reply pkg/auth/handlers.go]: actually, adding another log line.  the underlying function should log as well, but verbosity is good in this case.

File: pkg/auth/handler_utils.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+const (
+	ErrIdpClient errors.ErrorCode = "IDP_REQUEST_FAILED"
+)
+
+/*
+This struct represents what should be returned by an IDP according to the specification at
+ https://openid.net/specs/openid-connect-core-1_0.html#UserInfoResponse
+
+Keep in mind that not all fields are necessarily populated, and additional fields may be present as well. This is a sample
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply pkg/auth/handler_utils.go]: The OpenID Connect spec defines this response.  Not everyone sticks to it but we figured it was the best thing to use, even if there are some missing fields.

File: pkg/auth/cookie.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+const (
+	accessTokenCookie  = "flyte_jwt"
+	refreshTokenCookie = "flyte_refresh"
+	csrfStateCookie    = "flyte_csrf_state"
+	redirectUrlCookie  = "flyte_redirect_location"
+)
+
+const (
+	ErrSecureCookie errors.ErrorCode = "SECURE_COOKIE_ERROR"
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply pkg/auth/cookie.go]: already should be.  This error code is defined using the flytestdlib type and the error object that is created from it also uses that.

File: pkg/auth/auth_context.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	if options.IdpUserInfoEndpoint != "" {
+		base, err := url.Parse(options.BaseUrl)
+		if err != nil {
+			logger.Errorf(ctx, "Error parsing base URL %s", err)
+			return Context{}, errors.Wrapf(ErrAuthContext, err,
+				"Error parsing base URL while constructing IDP user info endpoint")
+		}
+		joinedPath := path.Join(base.EscapedPath(), options.IdpUserInfoEndpoint)
+
+		u, err := url.Parse(joinedPath)
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply pkg/auth/auth_context.go]: okay

File: pkg/auth/cookie.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+	csrfStateToken := NewCsrfToken(time.Now().UnixNano())
+	return http.Cookie{
+		Name:     csrfStateCookie,
+		Value:    csrfStateToken,
+		SameSite: http.SameSiteLaxMode,
+		HttpOnly: true,
+	}
+}
+
+func VerifyCsrfCookie(ctx context.Context, request *http.Request) bool {
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply pkg/auth/cookie.go]: changing to an error, removing log lines, and moving log to call site.

File: pkg/auth/handlers.go [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+func GetLoginHandler(ctx context.Context, authContext interfaces.AuthenticationContext) http.HandlerFunc {
+	return func(writer http.ResponseWriter, request *http.Request) {
+		csrfCookie := NewCsrfCookie()
+		csrfToken := csrfCookie.Value
+		http.SetCookie(writer, &csrfCookie)
+
+		state := HashCsrfState(csrfToken)
+		logger.Debugf(ctx, "Setting CSRF state cookie to %s and state to %s\n", csrfToken, state)
+		url := authContext.OAuth2Config().AuthCodeURL(state)
+		fmt.Println(url)
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply pkg/auth/handlers.go]: will keep until shortly before merge, then will remove

File: pkg/auth/handlers.go [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+		token, err := ParseAndValidate(ctx, authContext.Claims(), tokenStr, authContext.OidcProvider())
+		if err != nil {
+			return ctx, status.Errorf(codes.Unauthenticated, "could not parse token string into object: %s %s", tokenStr, err)
+		}
+		if token == nil {
+			return ctx, status.Errorf(codes.Unauthenticated, "Token was nil after parsing %s", tokenStr)
+		} else if token.Subject == "" {
+			return ctx, status.Errorf(codes.Unauthenticated, "no email or empty email found")
+		} else {
+			newCtx := WithUserEmail(context.WithValue(ctx, "bearer", tokenStr), token.Subject)
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply pkg/auth/handlers.go]: Can I punt on this and the above comment until the logging work?  I suspect all this will change.

**Comment by honnix** [comment]
As a reminder, https://github.com/lyft/flyteadmin/pull/8/files#diff-10cb7b7d10da61a9793a32f86a893485L5 makes sandbox deployment in `flyte` repo incompatible now with master.

**Comment by wild-endeavor** [comment]
@honnix thank you... I'll fix soon.  forgot to do that.

**Pull request merged**